### ISSUE TITLE
Fix S3242: Rule should not report when more general type doesn't have indexer

### ISSUE THREAD

**Issue by vitaliyzhukov** [SonarSource/sonar-csharp#640]
### Description

Rule S3242 triggers for types with indexer, base types of which don't have one. In case described below IList has indexer and ICollection doesn't.

### Repro steps

```
public class Foo
{
    public void MethodOne(IList list)
    {
        if (list.Count > 0)
        {
            Console.WriteLine(list[0]);
        }
    }

    public void MethodTwo(IList<Foo> list)
    {
        if (list.Count > 0)
        {
            Console.WriteLine(list[0]);
        }
    }
}
```

### Expected behavior

Rule should not trigger.

### Actual behavior

Rule triggers with following messages: 

For 'MethodOne':
> "Consider using more general type 'System.Collections.ICollection' instead of 'System.Collections.IList'."

For 'MethodTwo':
> "Consider using more general type 'System.Collections.Generic.ICollection<MyProject.Foo>' instead of 'System.Collections.Generic.IList<MyProject.Foo>'."

### Related information

* SonarC# Version: v6.2.0.2536
* Visual Studio Version: VS2015, VS 2017

**Comment by Evangelink** [comment]
@michalb-sonar I will let you handle this one when you get back from holiday.
